src/System.Text.JsonLab/System/Text/Json/JsonReader.cs

```diff
@@ -1471,21 +1365,34 @@ private void AdjustLineNumber(ReadOnlySpan<byte> span)
             }
             _position = span.Length + 1;
         }
-#else
+#endif
+
         // https://tools.ietf.org/html/rfc7159#section-6
-        //TODO: Investigate optimizations
-        private void ValidateNumber(ReadOnlySpan<byte> data)
+        private ReadOnlySpan<byte> GetNumber(ReadOnlySpan<byte> data)
         {
             Debug.Assert(data.Length > 0);
 
+            ReadOnlySpan<byte> delimiters = JsonConstants.Delimiters;
+
             int i = 0;
             byte nextByte = data[i++];
 
             if (nextByte == '-')
             {
                 if (i >= data.Length)
+#if UseInstrumented
+                    throw new JsonReaderException($"Invalid number. Last character read: '{(char)nextByte}'. Expected a digit.", _lineNumber, _position);
+#else
                     JsonThrowHelper.ThrowJsonReaderException(ref this);
+#endif
+
                 nextByte = data[i++];
+                if (delimiters.IndexOf(nextByte) != -1)
```

Comment:
This will do vectorized IndexOf, which we know is slower for very small buffers. How large is the delimiters buffer? Is it worth doing vectorized IndexOf here?

Reply:
The delimiters buffer is 7 elements: https://github.com/dotnet/corefxlab/blob/952ca00c3e5a6c930447cf53116f45b2aa007b67/src/System.Text.JsonLab/System/Text/Json/JsonConstants.cs#L44

I am not sure if its worth it to special case it.

There was no noticeable difference between a naiive for loop and using IndexOf here (within error).
Reading an array of 10 million numbers:
Before: 281.8 ms
After: 282.3 ms

The break-even point tends to be for buffer of size 3 (if we find the search item at the end). Otherwise, for size <= 3, a for-loop is 25% faster.


Method | DataLength | Mean | Scaled
-- | -- | -- | --
  |   |   |  
IndexOf | 1 | 317.2 ns | 1
ForLoop | 1 | 231.1 ns | 0.73
  |   |   |  
IndexOf | 2 | 424.0 ns | 1
ForLoop | 2 | 320.9 ns | 0.76
  |   |   |  
IndexOf | 3 | 526.6 ns | 1
ForLoop | 3 | 383.5 ns | 0.73
  |   |   |  
IndexOf | 4 | 346.8 ns | 1
ForLoop | 4 | 501.5 ns | 1.45
  |   |   |  
IndexOf | 5 | 378.5 ns | 1
ForLoop | 5 | 539.5 ns | 1.43
  |   |   |  
IndexOf | 6 | 539.2 ns | 1
ForLoop | 6 | 609.2 ns | 1.13
  |   |   |  
IndexOf | 7 | 560.6 ns | 1
ForLoop | 7 | 655.2 ns | 1.17
  |   |   |  
IndexOf | 8 | 366.0 ns | 1
ForLoop | 8 | 727.7 ns | 1.99
  |   |   |  
IndexOf | 9 | 451.2 ns | 1
ForLoop | 9 | 770.9 ns | 1.71
  |   |   |  
IndexOf | 10 | 513.6 ns | 1
ForLoop | 10 | 821.5 ns | 1.6
  |   |   |  
IndexOf | 15 | 678.8 ns | 1
ForLoop | 15 | 1,080.6 ns | 1.59
  |   |   |  
IndexOf | 16 | 568.7 ns | 1
ForLoop | 16 | 1,153.8 ns | 2.03
  |   |   |  
IndexOf | 17 | 646.0 ns | 1
ForLoop | 17 | 1,217.2 ns | 1.88
  |   |   |  
IndexOf | 20 | 682.7 ns | 1
ForLoop | 20 | 1,383.9 ns | 2.03
  |   |   |  
IndexOf | 21 | 726.0 ns | 1
ForLoop | 21 | 1,414.0 ns | 1.95



Calling IndexOf for size 7 is up to 15% slower to  up to15% faster depending on whether we find the search item (and where).


Method | FoundAt | Mean | Scaled
-- | -- | -- | --
IndexOf | 1 | 233.8 ns | 1
ForLoop | 1 | 197.1 ns | 0.84
  |   |   |  
IndexOf | 2 | 272.6 ns | 1
ForLoop | 2 | 322.0 ns | 1.18
  |   |   |  
IndexOf | 3 | 285.4 ns | 1
ForLoop | 3 | 406.1 ns | 1.42
  |   |   |  
IndexOf | 4 | 302.1 ns | 1
ForLoop | 4 | 481.6 ns | 1.59
  |   |   |  
IndexOf | 5 | 482.2 ns | 1
ForLoop | 5 | 511.0 ns | 1.06
  |   |   |  
IndexOf | 6 | 574.7 ns | 1
ForLoop | 6 | 553.0 ns | 0.96
  |   |   |  
IndexOf | 7 | 604.9 ns | 1
ForLoop | 7 | 572.3 ns | 0.95

```C#
[Benchmark(Baseline = true)]
public int IndexOf()
{
    ReadOnlySpan<byte> delimiters = Delimiters.AsSpan(0, DataLength);
    byte lookup = delimiters[DataLength - 1];
    int sum = 0;
    for (int i = 0; i < 100; i++)
        sum += delimiters.IndexOf(lookup);
    return sum;
}
```

Reply:
Note, if we expect not to find the lookup byte within the buffer, then the break-even point is 8 (not 3).

Search element not found:

Method | DataLength | Mean | Scaled
-- | -- | -- | --
  |   |   |  
IndexOf | 1 | 347.4 ns | 1
ForLoop | 1 | 202.4 ns | 0.58
  |   |   |  
IndexOf | 2 | 431.2 ns | 1
ForLoop | 2 | 234.9 ns | 0.55
  |   |   |  
IndexOf | 3 | 506.1 ns | 1
ForLoop | 3 | 354.3 ns | 0.7
  |   |   |  
IndexOf | 4 | 316.9 ns | 1
ForLoop | 4 | 437.9 ns | 1.38
  |   |   |  
IndexOf | 5 | 406.1 ns | 1
ForLoop | 5 | 373.2 ns | 0.92
  |   |   |  
IndexOf | 6 | 508.8 ns | 1
ForLoop | 6 | 396.9 ns | 0.78
  |   |   |  
IndexOf | 7 | 567.6 ns | 1
ForLoop | 7 | 420.2 ns | 0.74
  |   |   |  
IndexOf | 8 | 443.1 ns | 1
ForLoop | 8 | 446.4 ns | 1.01
  |   |   |  
IndexOf | 9 | 480.9 ns | 1
ForLoop | 9 | 488.9 ns | 1.02
  |   |   |  
IndexOf | 10 | 522.1 ns | 1
ForLoop | 10 | 532.5 ns | 1.02
  |   |   |  
IndexOf | 15 | 631.8 ns | 1
ForLoop | 15 | 687.5 ns | 1.09
  |   |   |  
IndexOf | 16 | 619.1 ns | 1
ForLoop | 16 | 731.9 ns | 1.18
  |   |   |  
IndexOf | 17 | 677.9 ns | 1
ForLoop | 17 | 779.4 ns | 1.15
  |   |   |  
IndexOf | 20 | 688.6 ns | 1
ForLoop | 20 | 837.6 ns | 1.22
  |   |   |  
IndexOf | 21 | 812.2 ns | 1
ForLoop | 21 | 1,083.0 ns | 1.33

Reply:
> This will do vectorized IndexOf, which we know is slower for very small buffers. How large is the delimiters buffer? Is it worth doing vectorized IndexOf here?

Also, note that I removed that specific use of IndexOf (and a few others). Explicitly looking for a digit was the correct behavior, rather than looking for a delimiter. So, we are using IndexOf sparingly now (essentially once before returning in the common case of integers, twice for decimals, and 3x if we have the e-notation).